project_key/models/project_project.py

```diff
@@ -0,0 +1,183 @@
+# Copyright 2017 - 2018 Modoolar <info@modoolar.com>
+# License LGPLv3.0 or later (https://www.gnu.org/licenses/lgpl-3.0.en.html).
+
+from odoo import models, fields, api, _
+from odoo.osv import expression
+
+
+class ProjectProject(models.Model):
+    _inherit = "project.project"
+
+    task_key_sequence_id = fields.Many2one(
+        comodel_name='ir.sequence',
+        string='Key Sequence',
+        ondelete="restrict",
+    )
+
+    key = fields.Char(
+        string='Key',
+        size=10,
+        required=True,
+        index=True,
+    )
+
+    _sql_constraints = [
+        ("project_key_unique", "UNIQUE(key)", "Project key must be unique!")
+    ]
+
+    @api.multi
+    @api.onchange('name')
+    def _onchange_project_name(self):
+        for rec in self:
+            if rec.name:
+                rec.key = self.generate_project_key(rec.name)
+            else:
+                rec.key = ''
+
+    @api.model
+    def create(self, vals):
+        if 'key' not in vals:
+            if vals['name']:
+                vals['key'] = self.generate_project_key(vals['name'])
+            else:
+                vals['key'] = ''
+
+        new_project = super(ProjectProject, self).create(vals)
+        new_project.create_sequence()
+
+        return new_project
+
+    @api.multi
+    def write(self, values):
+
+        reindex = False
+
+        if 'key' in values:
+            key = values['key']
+            reindex = self.key != key
+
+        res = super(ProjectProject, self).write(values)
+
+        if reindex:
+            # Here we don't expect to have more than one record
+            # because we can not have multiple projects with same KEY.
+            self.update_sequence()
+            self._reindex_task_keys()
+
+        return res
+
+    @api.multi
+    def unlink(self):
+        for project in self:
+            project.task_key_sequence_id.sudo().unlink()
+        return super(ProjectProject, self).unlink()
+
+    @api.model
+    def name_search(self, name, args=None, operator='ilike', limit=100):
+        args = args or []
+        domain = []
+        if name:
+            domain = [
+                '|', ('key', 'ilike', name + '%'), ('name', operator, name)
+            ]
+            if operator in expression.NEGATIVE_TERM_OPERATORS:
+                domain = ['&'] + domain
```

Comment:
Overriding the domain in `name_search` is always tricky. For instance, if args is not defined, it looks like you end up with an unbalanced domain. The same can happen if args is an unbalanced domain itself. And to be compatible with any other module that might be installed on the same database, it would be better to call the super method. But my main question is, what is the goal of this method override?

Reply:
The idea behind this method was to include task ``code`` into name search.

Reply:
Am I missing something? I don't see any references to task `code` in this method.

Reply:
To restore the compatibility with other modules (e.g. core project and analytic modules which have their own name_get logic), I would reap a list of ids from super and do a new search on [('|', ('key', 'ilike', name + '%'), ('id', 'in', ids)] and return name_get on the result of that.

Reply:
Fixed.